File: solutions/go/react/9/react.go
package react

import "slices"

type canceler struct {
    cell *cell
    callback *func(int)
}

type cell struct {
    value int
    computation *func(deps ...int) int
    deps []*cell
    callbacks []*func(int)
    reactor *reactor
}

type reactor map[*cell][]*cell

func (c *canceler) Cancel() {
    c.cell.callbacks = slices.DeleteFunc(c.cell.callbacks, func(cb *func(int)) bool {
        return c.callback == cb
    })
}

func (c *cell) Value() int {
	if len(c.deps) == 0 {
        return c.value
    }
    depValueList := make([]int, len(c.deps))
    for i, dep := range(c.deps) {
        depValueList[i] = dep.Value()
    }
    return (*c.computation)(depValueList...)
}

func (c *cell) SetValue(value int) {
	affectedIsCheckedMap := map[*cell]bool{c: false}
    for {
        addedCountPerCycle := 0
        for affectedCell, isChecked := range affectedIsCheckedMap {
            if isChecked {
                continue
            }
            for _, affectedByAffected := range (*c.reactor)[affectedCell] {
                _, alreadyAdded := affectedIsCheckedMap[affectedByAffected]
                if alreadyAdded {
                    continue
                }
                affectedIsCheckedMap[affectedByAffected] = false
                addedCountPerCycle++
            }
            affectedIsCheckedMap[affectedCell] = true
        }
        if addedCountPerCycle == 0 {
            break
        }
    }

    affectedCellPrevValueMap := make(map[*cell]int)
    for affectedCell := range affectedIsCheckedMap {
        affectedCellPrevValueMap[affectedCell] = affectedCell.Value()
    }

    c.value = value

    for affectedCell := range affectedIsCheckedMap {
        if affectedCell.Value() != affectedCellPrevValueMap[affectedCell] {
            affectedCellNewValue := affectedCell.Value()
            for _, affectedCellCallback := range affectedCell.callbacks {
                (*affectedCellCallback)(affectedCellNewValue)
            }
        }
    }
}

func (c *cell) AddCallback(callback func(int)) Canceler {
	c.callbacks = append(c.callbacks, &callback)
    var cancelerInstance canceler
    cancelerInstance.cell = c
    cancelerInstance.callback = &callback
    return &cancelerInstance
}

func New() Reactor {
    return &reactor{}
}

func (r *reactor) CreateInput(initial int) InputCell {
	var newCell cell
    newCell.value = initial
    newCell.deps = make([]*cell, 0)
    newCell.callbacks = make([]*func(int), 0)
    newCell.reactor = r
    return &newCell
}

func (r *reactor) CreateCompute1(dep Cell, compute func(int) int) ComputeCell {
	var newCell cell
    newCell.deps = make([]*cell, 1)
    newCell.deps[0] = dep.(*cell)
    variadicCompute := func (deps ...int) int {
        if len(deps) != 1 {
            panic("Unexpected number of varargs. 1 expected.")
        }
        return compute(deps[0])
    }
    newCell.computation = &variadicCompute
    newCell.callbacks = make([]*func(int), 0)
    newCell.reactor = r

    for _, parent := range newCell.deps {
        siblings, siblingExists := (*r)[parent]
        if !siblingExists {
            (*r)[parent] = make([]*cell, 0)
        }
        (*r)[parent] = append(siblings, &newCell)
    }
    return &newCell
}

func (r *reactor) CreateCompute2(dep1, dep2 Cell, compute func(int, int) int) ComputeCell {
	var newCell cell
    newCell.deps = make([]*cell, 2)
    newCell.deps[0] = dep1.(*cell)
    newCell.deps[1] = dep2.(*cell)
    variadicCompute := func (deps ...int) int {
        if len(deps) != 2 {
            panic("Unexpected number of varargs. 2 expected.")
        }
        return compute(deps[0], deps[1])
    }
    newCell.computation = &variadicCompute
    newCell.callbacks = make([]*func(int), 0)
    newCell.reactor = r

    for _, parent := range newCell.deps {
        siblings, siblingExists:= (*r)[parent]
        if !siblingExists {
            (*r)[parent] = make([]*cell, 0)
        }
        (*r)[parent] = append(siblings, &newCell)
    } 
    return &newCell
}

File: solutions/lua/react/9/react.lua
local function reactor()
	local update_fns <const> = {}

	local function update_after(index)
		for i = index + 1, #update_fns do
			update_fns[i]()
		end
	end

	local function input_cell(initial_value)
		local value = initial_value
		local index <const> = #update_fns

		local function get_value()
			return value
		end

		local function set_value(new_value)
			if value == new_value then
				return
			end
			value = new_value
			update_after(index)
		end

		return { get_value = get_value, set_value = set_value }
	end

	local function compute_cell(...)
		local args <const> = { ... }
		local fn <const> = table.remove(args)
		local arg_vals <const> = {}

		local function compute()
			for i, cell in ipairs(args) do
				arg_vals[i] = cell.get_value()
			end
			return fn(table.unpack(arg_vals))
		end

		local value = compute()
		local callbacks <const> = {}

		table.insert(update_fns, function()
			local new_value <const> = compute()
			if value == new_value then
				return
			end
			value = new_value
			for cb, _ in pairs(callbacks) do
				cb(value)
			end
		end)

		local function get_value()
			return value
		end

		local function watch(callback)
			callbacks[callback] = true
		end

		local function unwatch(callback)
			callbacks[callback] = nil
		end

		return { get_value = get_value, watch = watch, unwatch = unwatch }
	end

	return { InputCell = input_cell, ComputeCell = compute_cell }
end

return { Reactor = reactor }
